File: packages/provide-uterm-go/vt/sgr.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package vt

import "fmt"

// SelectGraphicRendition sets display attributes on the cursor (SGR).
// Recognized codes match pyte exactly: reset (0), bold/italics/underscore/
// blink/reverse/strikethrough and their resets, 16-color ANSI plus aixterm
// bright colors, and extended 256-color (38;5;N / 48;5;N) and truecolor
// (38;2;R;G;B / 48;2;R;G;B) selections. Unknown codes are ignored.
func (s *Screen) SelectGraphicRendition(attrs ...int) {
	// Fast path for resetting all attributes.
	if len(attrs) == 0 || (len(attrs) == 1 && attrs[0] == 0) {
		s.cursor.Attrs = s.defaultChar()
		return
	}

	work := s.cursor.Attrs
	i := 0
	for i < len(attrs) {
		attr := attrs[i]
		i++
		switch {
		case attr == 0:
			// Reset all attributes.
			work = s.defaultChar()
		case fgANSI[attr] != "":
			work.FG = fgANSI[attr]
		case bgANSI[attr] != "":
			work.BG = bgANSI[attr]
		case applyTextAttr(&work, attr):
			// Text style flag applied.
		case fgAIXTerm[attr] != "":
			work.FG = fgAIXTerm[attr]
		case bgAIXTerm[attr] != "":
			work.BG = bgAIXTerm[attr]
		case attr == 38 || attr == 48:
			i = applyExtendedColor(&work, attr == 38, attrs, i)
		}
	}

	s.cursor.Attrs = work
}

// applyTextAttr applies an SGR text style code to c, reporting whether the
// code was a known style code (pyte.graphics.TEXT).
func applyTextAttr(c *Char, attr int) bool {
	switch attr {
	case 1:
		c.Bold = true
	case 3:
		c.Italics = true
	case 4:
		c.Underscore = true
	case 5:
		c.Blink = true
	case 7:
		c.Reverse = true
	case 9:
		c.Strikethrough = true
	case 22:
		c.Bold = false
	case 23:
		c.Italics = false
	case 24:
		c.Underscore = false
	case 25:
		c.Blink = false
	case 27:
		c.Reverse = false
	case 29:
		c.Strikethrough = false
	default:
		return false
	}
	return true
}

// applyExtendedColor handles the parameters following SGR 38/48 starting
// at attrs[i], returning the next unconsumed index. Out-of-range palette
// indices and truncated parameter lists consume their parameters without
// effect, matching pyte's IndexError swallowing.
func applyExtendedColor(c *Char, isFG bool, attrs []int, i int) int {
	if i >= len(attrs) {
		return i
	}
	n := attrs[i]
	i++
	switch n {
	case 5: // 256-color palette.
		if i >= len(attrs) {
			return i
		}
		m := attrs[i]
		i++
		if m < len(fgBG256) {
			setColor(c, isFG, fgBG256[m])
		}
	case 2: // 24-bit color.
		if i+3 > len(attrs) {
			return len(attrs)
		}
		setColor(c, isFG, fmt.Sprintf("%02x%02x%02x", attrs[i], attrs[i+1], attrs[i+2]))
		i += 3
	}
	return i
}

func setColor(c *Char, isFG bool, color string) {
	if isFG {
		c.FG = color
	} else {
		c.BG = color
	}
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Vt/Sgr.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Vt;

public partial class Screen
{
    public void SelectGraphicRendition(params int[] attrs)
    {
        if (attrs.Length == 0 || (attrs.Length == 1 && attrs[0] == 0))
        {
            _cursor.Attrs = DefaultChar();
            return;
        }

        var work = _cursor.Attrs;
        var i = 0;
        while (i < attrs.Length)
        {
            var attr = attrs[i++];
            if (attr == 0)
            {
                work = DefaultChar();
            }
            else if (Graphics.FgAnsi.TryGetValue(attr, out var fg))
            {
                work.FG = fg;
            }
            else if (Graphics.BgAnsi.TryGetValue(attr, out var bg))
            {
                work.BG = bg;
            }
            else if (ApplyTextAttr(ref work, attr))
            {
                // style
            }
            else if (Graphics.FgAixTerm.TryGetValue(attr, out var afg))
            {
                work.FG = afg;
            }
            else if (Graphics.BgAixTerm.TryGetValue(attr, out var abg))
            {
                work.BG = abg;
            }
            else if (attr is 38 or 48)
            {
                i = ApplyExtendedColor(ref work, attr == 38, attrs, i);
            }
        }

        _cursor.Attrs = work;
    }

    private static bool ApplyTextAttr(ref Char c, int attr)
    {
        switch (attr)
        {
            case 1: c.Bold = true; break;
            case 3: c.Italics = true; break;
            case 4: c.Underscore = true; break;
            case 5: c.Blink = true; break;
            case 7: c.Reverse = true; break;
            case 9: c.Strikethrough = true; break;
            case 22: c.Bold = false; break;
            case 23: c.Italics = false; break;
            case 24: c.Underscore = false; break;
            case 25: c.Blink = false; break;
            case 27: c.Reverse = false; break;
            case 29: c.Strikethrough = false; break;
            default: return false;
        }

        return true;
    }

    private static int ApplyExtendedColor(ref Char c, bool isFg, int[] attrs, int i)
    {
        if (i >= attrs.Length)
        {
            return i;
        }

        var n = attrs[i++];
        switch (n)
        {
            case 5:
                if (i >= attrs.Length)
                {
                    return i;
                }

                var m = attrs[i++];
                if (m < Graphics.FgBg256.Length)
                {
                    SetColor(ref c, isFg, Graphics.FgBg256[m]);
                }

                break;
            case 2:
                if (i + 3 > attrs.Length)
                {
                    return attrs.Length;
                }

                SetColor(ref c, isFg, $"{attrs[i]:x2}{attrs[i + 1]:x2}{attrs[i + 2]:x2}");
                i += 3;
                break;
        }

        return i;
    }

    private static void SetColor(ref Char c, bool isFg, string color)
    {
        if (isFg)
        {
            c.FG = color;
        }
        else
        {
            c.BG = color;
        }
    }
}
